File: crypto/src/main/java/net/corda/v5/crypto/merkle/IndexedMerkleLeaf.java
package net.corda.v5.crypto.merkle;

import net.corda.v5.base.annotations.CordaSerializable;
import net.corda.v5.base.annotations.DoNotImplement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * {@link IndexedMerkleLeaf}s are building blocks of {@link MerkleProof}s.
 * They contain the required information about a particular leaf which is needed for the verification.
 */
@DoNotImplement
@CordaSerializable
public interface IndexedMerkleLeaf {

    /**
     * Obtain the index of this leaf.
     *
     * @return Integer leaf index.
     */
    public int getIndex();

    /**
     * Obtain the nonce of this tree leaf.
     *
     * @return Nonce as a byte array.
     */
    @Nullable
    public byte[] getNonce();

    /**
     * Obtain the data for this leaf.
     *
     * @return Leaf data as a byte array.
     */
    @NotNull
    public byte[] getLeafData();
}

File: crypto/src/main/kotlin/net/corda/v5/crypto/merkle/IndexedMerkleLeaf.kt
package net.corda.v5.crypto.merkle

import net.corda.v5.base.annotations.CordaSerializable

@CordaSerializable
class IndexedMerkleLeaf(
    val index: Int,
    val nonce: ByteArray?,
    val leafData: ByteArray
) {
    override fun toString(): String {
        return "Leaf($index)[${leafData.size} bytes]"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as IndexedMerkleLeaf

        if (index != other.index) return false
        if (nonce != null) {
            if (other.nonce == null) return false
            if (!nonce.contentEquals(other.nonce)) return false
        } else if (other.nonce != null) return false
        if (!leafData.contentEquals(other.leafData)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = index
        result = 31 * result + (nonce?.contentHashCode() ?: 0)
        result = 31 * result + leafData.contentHashCode()
        return result
    }

}